### PR TITLE
add progress bar support for apptainer push/pull for oras protocol

### LINKED ISSUE #437: Feature request: progress indicator during `apptainer push`
Apptainer should support an optional progress indicator during push, with percentage complete as well as bytes transferred / total bytes to transfer.

### PR DESCRIPTION
**Pull request by JasonYangShadow**
## Description of the Pull Request (PR):

add progress bar support for `apptainer push/pull` for oras protocol

### This fixes or addresses the following GitHub issues:

 - Fixes #437 

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### Apptainer Push
push

### Apptainer Pull
pull

### CODE REVIEW

**Comment by JasonYangShadow** [comment]
I will also add progress bar to `apptainer pull oras://` as well. Please do not merge until the code is applied.

**Comment by panda1100** [comment]
I tested push/pull and both works great!

**Comment by DrDaveD** [comment]
I'm surprised this has a whole new source file just for the oras protocol.  Aren't there progress bars on other protocols that can share code?  Or is this the first one and it should be extended to others?  Why should we treat the oras protocol specially?

**Comment by panda1100** [comment]
@DrDaveD -san,
This is not the first progress bar implementation. `docker://` pull and `library://` push/pull has progress bar. Each has its own implementation. 
The reason we need unique implementation for oras "as well", as my understanding from discussing with @JasonYangShadow and @cclerget about this PR design, we need to access reader/writer buffer to track progress to implement progress bar, to access those buffer we need to control over the reader/writer returned by Push/Fetch method those are under oras library (those are not exposed as callable function?). oras library itself doesn't have progress bar implementation yet. so, we need to use those interfaces directly from Apptainer. This is the reason we decided to implement this way.

**Comment by panda1100** [comment]
Thank you @DrDaveD -san, I see your point. I'll discuss with @JasonYangShadow about it tomorrow morning.

**Comment by JasonYangShadow** [comment]
I will take a look at current progressbar and mpb related implementation and see if we can reuse mpb.

**Comment by JasonYangShadow** [comment]
@panda1100 I have rewritten the function using apptainer existing progressbar, the same we used previously. Please help verify and check if it is okay.

**Comment by panda1100** [comment]
@JasonYangShadow Thank you Jason! I did test on my end and works good for both push and pull 👍

**Review by DrDaveD — CHANGES_REQUESTED** [review]
This is an improvement.  However I still don't like that there's a new function called `getOrasResolver()` along with an old function called `getResolver()` with the same parameters and only for oras, in fact in oras.go.  It also looks like they're mostly identical code too.

I suggest renaming `getOrasResolver` to `getResolver` and adding a flag that makes it optionally behave like the existing `getResolver`, then remove the current `getResolver`.

**Comment by JasonYangShadow** [comment]
> This is an improvement. However I still don't like that there's a new function called `getOrasResolver()` along with an old function called `getResolver()` with the same parameters and only for oras, in fact in oras.go. It also looks like they're mostly identical code too.
> 
> I suggest renaming `getOrasResolver` to `getResolver` and adding a flag that makes it optionally behave like the existing `getResolver`, then remove the current `getResolver`.

Thanks for the comments, next time I will be more serious on the duplicated code.

File: internal/pkg/client/oras/resolver.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
+package oras
```
Comment (DrDaveD) [inline internal/pkg/client/oras/resolver.go:1]: Please include the standard header on this file, and also include a comment to say what the file is for, to add progress bars.

File: internal/pkg/client/oras/oras.go [hunk 2]
```diff
@@ -86,7 +86,7 @@ func newOrasUploadTransport() http.RoundTripper {
 	}
 }
 
-func getResolver(ctx context.Context, ociAuth *ocitypes.DockerAuthConfig, noHTTPS, push bool) (remotes.Resolver, error) {
+func getResolver(ctx context.Context, ociAuth *ocitypes.DockerAuthConfig, noHTTPS, push, orasResolver bool) (remotes.Resolver, error) {
```
Comment (DrDaveD) [inline internal/pkg/client/oras/oras.go:89]: I don't think the name `orasResolver` is very descriptive.  How about something like `progressBar` or `includeProgressBar`.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

File: internal/pkg/client/oras/oras.go (hunk 2 above)
Comment (JasonYangShadow) [reply internal/pkg/client/oras/oras.go]: renamed to `progressBar` thanks!
https://github.com/apptainer/apptainer/pull/1587/commits/aab871d7c4c217c169908a6e6afddc9f49654c3f

File: internal/pkg/client/oras/resolver.go (hunk 1 above)
Comment (JasonYangShadow) [reply internal/pkg/client/oras/resolver.go:12]: thanks! added
https://github.com/apptainer/apptainer/pull/1587/commits/aab871d7c4c217c169908a6e6afddc9f49654c3f

**Review by DrDaveD — APPROVED** [review]
Please also cherry-pick to release-1.2

**Comment by panda1100** [comment]
Thank you DrDave-san and Jason!

**Pull request merged**